File: python/cps_utility_console/validators.py
"""
CPS Utility Console - Input Validators
Author: Dominik Żurek
"""


def get_float(prompt: str) -> float:
    """
    Prompt user for a float value.
    Keeps asking until valid input is provided.
    """
    while True:
        raw = input(prompt).strip()
        try:
            return float(raw)
        except ValueError:
            print(f"  Error: '{raw}' is not a valid number. Try again.")


def get_float_in_range(prompt: str, min_val: float, max_val: float) -> float:
    """
    Prompt user for a float value within [min_val, max_val].
    Keeps asking until valid input is provided.
    """
    while True:
        value = get_float(prompt)
        if min_val <= value <= max_val:
            return value
        print(f"  Error: value must be between {min_val} and {max_val}. Try again.")


def get_positive_int(prompt: str) -> int:
    """
    Prompt user for a positive integer (>= 1).
    Keeps asking until valid input is provided.
    """
    while True:
        raw = input(prompt).strip()
        try:
            value = int(raw)
            if value >= 1:
                return value
            print(f"  Error: value must be a positive integer. Try again.")
        except ValueError:
            print(f"  Error: '{raw}' is not a valid integer. Try again.")


def get_choice(prompt: str, valid: list[str]) -> str:
    """
    Prompt user to select from a list of valid string choices.
    Keeps asking until valid input is provided.
    """
    while True:
        raw = input(prompt).strip().upper()
        if raw in valid:
            return raw
        print(f"  Error: choose one of {valid}. Try again.")

File: cpp/src/validators.cpp
/*
 * CPS Utility Console - Input Validators
 * Author: Dominik Żurek
 */

#include "../include/validators.hpp"
#include <iostream>
#include <limits>
#include <algorithm>
#include <cctype>

float cps::getFloat(const std::string& prompt) {
    float value;
    while (true) {
        std::cout << prompt;
        if (std::cin >> value) {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            return value;
        }
        std::cout << "  Error: invalid number. Try again.\n";
        std::cin.clear();
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    }
}

float cps::getFloatInRange(const std::string& prompt, float min, float max) {
    while (true) {
        float value = cps::getFloat(prompt);
        if (value >= min && value <= max) return value;
        std::cout << "  Error: value must be between " << min << " and " << max << ". Try again.\n";
    }
}

int cps::getPositiveInt(const std::string& prompt) {
    while (true) {
        std::cout << prompt;
        int value;
        if (std::cin >> value) {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            if (value >= 1) return value;
            std::cout << "  Error: value must be a positive integer. Try again.\n";
        } else {
            std::cout << "  Error: not a valid integer. Try again.\n";
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        }
    }
}

std::string cps::getChoice(const std::string& prompt, const std::vector<std::string>& valid) {
    while (true) {
        std::cout << prompt;
        std::string input;
        std::getline(std::cin, input);
        for (char& c : input) c = toupper(c);
        if (std::find(valid.begin(), valid.end(), input) != valid.end()) return input;
        std::cout << "  Error: invalid choice. Try again.\n";
    }
}